src/Raven.Server/Documents/TransactionOperationsMerger.cs

```diff
@@ -848,5 +1125,283 @@ public void Dispose()
                 result.TaskCompletionSource.TrySetCanceled();
             }
         }
+
+
+        public enum TxInstruction
+        {
+            BeginTx,
+            Commit,
+            Rollback,
+            DisposeTx,
+            BeginAsyncCommitAndStartNewTransaction,
+            EndAsyncCommit,
+            DisposePrevTx
+        }
+
+        private RecordingState _recordingState = null;
+
+        public void StartRecording(string filePath)
+        {
+            if (_recordingState != null)
+                return;
+            Interlocked.CompareExchange(ref _recordingState, new RecordingState.BeforeEnabledRecordingState(this, filePath), null);
+            _waitHandle.Set();
+        }
+
+        public void StopRecording()
+        {
+            var recordingState = _recordingState;
+            if (recordingState != null)
+            {
+                recordingState.Shutdown();
+                _waitHandle.Set();
+            }
+        }
+
+        public IEnumerable<ReplayProgress> Replay(Stream replayStream)
+        {
```

Comment:
Can we move the implementation to some dedicated helper class?

Reply:
done